### PR TITLE
Reduce queries for products in checkout views

### LINKED ISSUE #603: Review checkout and optimize number of SQL queries
New decorator-based approach in checkout introduced some additional SQL queries, we should test the checkout process and add `prefetch_related` / `select_related` queries when it's required.

### PR DESCRIPTION
**Pull request by krzysztofwolski**
Fix #603

### CODE REVIEW

File: saleor/checkout/core.py [hunk 1]
```diff
@@ -28,6 +29,8 @@ def __init__(self, cart, user, tracking_code):
         self.tracking_code = tracking_code
         self.user = user
         self.discounts = cart.discounts
+        self.cached_shipping_method = None
```
Comment (maciekjablonski) [inline saleor/checkout/core.py]: Maybe:
https://docs.djangoproject.com/en/1.10/ref/utils/#django.utils.functional.cached_property ?

Comment (krzysztofwolski) [inline saleor/checkout/core.py]: @maciekjablonski Tried it, but cached_property does not work with property setters.

Comment (artursmet) [inline saleor/checkout/core.py]: I'm not sure if `cached_` prefix object fits here best. This cache is an implementation detail and it won't be never exposed as an public part of Checkout object API, so I recommend to rename these parameters to private fields like `self._shipping_method_cache` and `self._shipping_address_cache`

File: saleor/order/views.py [hunk 2]
```diff
@@ -18,15 +18,21 @@
...
     order = get_object_or_404(orders, token=token)
     groups = order.groups.all()
     return TemplateResponse(request, 'order/details.html',
                             {'order': order, 'groups': groups})
 
 
 def payment(request, token):
-    orders = Order.objects.prefetch_related('groups__items')
+    orders = Order.objects.prefetch_related('groups__items',
+                                            'groups__items__product')
```
Comment (patrys) [inline saleor/order/views.py]: Doesn't `groups__items__product` imply `groups__items`?

**Comment by krzysztofwolski** [comment]
@patrys Fixed!

File: saleor/checkout/core.py [hunk 3]
```diff
@@ -83,9 +86,12 @@ def deliveries(self):
 
     @property
     def shipping_address(self):
+        if self._shipping_address is not None:
```
Comment (patrys) [inline saleor/checkout/core.py]: I think a more common idiom would be:
```python
if self._shipping_address is None:
    # calculate it
    ...
return self._shipping_address
```

**Comment by krzysztofwolski** [comment]
@patrys Added cache checking in existing tests and fixed setters

**Comment by krzysztofwolski** [comment]
@patrys Is there anything else to fix?

**Comment by krzysztofwolski** [comment]
Rebased to current master and added new fixes.

**Pull request merged**